pkg/controller/clusterserviceversion/clusterserviceversion_controller.go

```diff
@@ -131,16 +164,65 @@ func (r *ReconcileClusterServiceVersion) Reconcile(request reconcile.Request) (r
 		return reconcile.Result{}, err
 	}
 
+	annotations := CSV.GetAnnotations()
+	if annotations == nil {
+		annotations = make(map[string]string)
+	}
+
+	//check if the CSV name, matches the RHM-Operator CSV name
+	if strings.Contains(CSV.GetName(), utils.CSV_NAME) {
+		// examine DeletionTimestamp to determine if object is under deletion
+		if CSV.ObjectMeta.DeletionTimestamp.IsZero() {
+			// Case 1: the object is not being deleted
+
+			// Check if there is a finalizer attached to CSV
+			if !utils.Contains(CSV.GetFinalizers(), utils.CSV_FINALIZER) {
+				// if no -> register finalizer
+				CSV.ObjectMeta.Finalizers = append(CSV.ObjectMeta.Finalizers, utils.CSV_FINALIZER)
+				if err := r.client.Update(context.Background(), CSV); err != nil {
+					return reconcile.Result{}, err
+				}
+				reqLogger.Info("added finailzer to CSV")
+			}
+
+			err = r.checkAnnotations(CSV, annotations)
+			if err != nil {
+				return reconcile.Result{}, err
+			}
+
+		} else {
+			// Case 2: The object is being deleted
+			if utils.Contains(CSV.GetFinalizers(), utils.CSV_FINALIZER) {
+				// our finalizer is present, so lets handle any external dependency
+				reqLogger.Info("deleting csv")
+				if err := r.deleteExternalResources(CSV); err != nil {
+					// if fail to delete the external dependency here, return with error
+					// so that it can be retried
+					reqLogger.Error(err, "unable to delete csv")
+					return reconcile.Result{}, err
+				}
+				// remove our finalizer from the list and update it.
+				CSV.SetFinalizers(utils.RemoveKey(CSV.GetFinalizers(), utils.CSV_FINALIZER))
+				if err := r.client.Update(context.Background(), CSV); err != nil {
+					return reconcile.Result{}, err
+				}
```

Comment:
Add this to a finalizer function, also please don't requeue on error.

Reply:
Double checking:
if error while deleting -> don't re-queue (may cause object to be stuck terminating)
if no error while deleting -> don't re-queue as object is terminating anyways